packages/test-sync/src/commonTest/kotlin/io/realm/kotlin/test/mongodb/common/GeoSpatialTests.kt

```diff
@@ -0,0 +1,229 @@
+
+        val config =
+            SyncConfiguration.Builder(
+                user = user,
+                schema = GEO_SCHEMA
+            ).initialSubscriptions(rerunOnOpen = true) {
+            }.build()
+
+        Realm.open(config).use {
+            it.writeBlocking {
+                assertFails {
+                    copyToRealm(SyncRestaurant())
+                }
+            }
+        }
+    }
+
+    @Test
+    fun geoBox_tests() {
+        generic_geo_test(
+            bounds = GeoBox.create(
+                top = 1.0, left = 1.0,
+                bottom = -1.0, right = -1.0,
+            ),
+            validLocation = Location(0.0, 0.0),
+            invalidLocation = Location(100.0, 100.0),
+        )
+    }
+
+    @Test
+    fun geoCircle_tests() {
+        generic_geo_test(
+            bounds = GeoCircle.create(
+                GeoPoint.create(0.0, 0.0), Distance.fromKilometers(.01)
+            ),
+            validLocation = Location(0.0, 0.0),
+            invalidLocation = Location(101.0, 101.0),
+        )
+    }
+
+    @Test
+    fun geoPolygon_tests() {
+        generic_geo_test(
+            bounds = GeoPolygon.create(
+                outerRing = listOf(
+                    GeoPoint.create(-5.0, -5.0),
+                    GeoPoint.create(5.0, -5.0),
+                    GeoPoint.create(5.0, 5.0),
+                    GeoPoint.create(-5.0, 5.0),
+                    GeoPoint.create(-5.0, -5.0)
+                ),
+                holes = arrayOf(
+                    listOf(
+                        GeoPoint.create(-4.0, -4.0),
+                        GeoPoint.create(4.0, -4.0),
+                        GeoPoint.create(4.0, 4.0),
+                        GeoPoint.create(-4.0, 4.0),
+                        GeoPoint.create(-4.0, -4.0)
+                    )
+                )
+            ),
+            validLocation = Location(4.5, 4.5), // Outside the hole and withing the ring
+            invalidLocation = Location(0.0, 0.0), // Inside the hole
+        )
+    }
+
+    private fun generic_geo_test(
+        bounds: Any,
+        validLocation: Location,
+        invalidLocation: Location,
+    ) {
+        val section = ObjectId()
+
+        // User #1 will try to write some data and assert some failure conditions.
+        val user1 = createRandomUser()
+        val config =
+            SyncConfiguration.Builder(
+                user = user1,
+                schema = GEO_SCHEMA
+            ).initialSubscriptions(rerunOnOpen = true) {
+                add(
+                    it.query<SyncRestaurant>(
+                        "section = $0 AND location GEOWITHIN $1",
+                        section,
+                        bounds
+                    )
+                )
+            }.waitForInitialRemoteData().build()
+
+        Realm.open(config).use { realm ->
+            runBlocking {
+                realm.subscriptions.waitForSynchronization(timeout = 30.seconds)
```

Comment:
This shouldn't be needed if you have `waitForInitialRemoteData`. It should take into account initial subscriptions.